KAFKA-18028: the effective kraft version of --no-initial-controllers should be 1 rather than 0

FrankYang0529 opened the pull request:
The `hasDynamicQuorum` only considers the `initialControllers`, which is configured when either `--initial-controllers` or `--standalone` is specified. It will result in the following unexpected behaviors:

1. Using the `no-initial-controllers` configuration get `kraft.version=0`
2. This leads to a command error when executing `--no-initial-controllers --feature kraft.version=1`

```
> ./bin/kafka-storage.sh format -t "zuV6tFqAQhSGfFe97O-9cw" --config ./config/kraft/controller3.properties --no-initial-controllers --feature kraft.version=1
Cannot set kraft.version to 1 unless KIP-853 configuration is present. Try removing the --feature flag for kraft.version.
```

To fix this, we should also allow `--no-initial-controllers` with `--feature kraft.version=1`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ahuang98 commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -143,6 +143,9 @@ object StorageTool extends Logging {
     }
+    if (namespace.getBoolean("no_initial_controllers")) {
+      formatter.setNoInitialControllers(true)
+    }
```
nit: can we change the if statement below into an else clause?

ahuang98 commented on metadata/src/main/java/org/apache/kafka/metadata/storage/Formatter.java:
```diff
@@ -130,6 +132,7 @@ public class Formatter {
      * The initial KIP-853 voters.
      */
     private Optional<DynamicVoters> initialControllers = Optional.empty();
+    private boolean noInitialControllers = false;
```
nit: thoughts on calling this `noInitialControllersFlag`? might make it more clear this is not just set when `initial-controllers` is not supplied

ahuang98 reviewed (changes requested):
thanks for addressing this issue!

FrankYang0529 commented:
Hi @ahuang98, thanks for the review. I address all comments.

FrankYang0529 commented:
Hi @ahuang98, thanks for the review. I will update description and test manually tomorrow.

FrankYang0529 commented:
Hi @ahuang98, I updated PR description and following is manually test results:

1. Setup three controller properties.

```
# path: ./config/kraft/controller1.properties
process.roles=controller
node.id=1
controller.quorum.bootstrap.servers=localhost:9093
listeners=CONTROLLER://:9093
advertised.listeners=CONTROLLER://localhost:9093
controller.listener.names=CONTROLLER
log.dirs=/tmp/kraft-controller-logs-1
```

```
# path: ./config/kraft/controller2.properties
process.roles=controller
node.id=2
controller.quorum.bootstrap.servers=localhost:9093
listeners=CONTROLLER://:9094
advertised.listeners=CONTROLLER://localhost:9094
controller.listener.names=CONTROLLER
log.dirs=/tmp/kraft-controller-logs-2
```

```
# path: ./config/kraft/controller3.properties
process.roles=controller
node.id=3
controller.quorum.bootstrap.servers=localhost:9093
listeners=CONTROLLER://:9095
advertised.listeners=CONTROLLER://localhost:9095
controller.listener.names=CONTROLLER
log.dirs=/tmp/kraft-controller-logs-3
```

2. Format node 1 as first standalone controller and start it.

```
> ./bin/kafka-storage.sh format -t "zuV6tFqAQhSGfFe97O-9cw" --config ./config/kraft/controller1.properties --standalone
Formatting dynamic metadata voter directory /tmp/kraft-controller-logs-1 with metadata.version 4.0-IV0.
> ./bin/kafka-server-start.sh ./config/kraft/controller1.properties
```

3. Format node 2 with no initial controller and start it.

```
> ./bin/kafka-storage.sh format -t "zuV6tFqAQhSGfFe97O-9cw" --config ./config/kraft/controller2.properties --no-initial-controllers --feature kraft.version=1
Formatting metadata directory /tmp/kraft-controller-logs-2 with metadata.version 4.0-IV0.
> ./bin/kafka-server-start.sh ./config/kraft/controller2.properties
```

4. Format node 3 with no initial controller and start it.

```
> ./bin/kafka-storage.sh format -t "zuV6tFqAQhSGfFe97O-9cw" --config ./config/kraft/controller3.properties --no-initial-controllers --feature kraft.version=1
Formatting metadata directory /tmp/kraft-controller-logs-3 with metadata.version 4.0-IV0.
> ./bin/kafka-server-start.sh ./config/kraft/controller3.properties
```

5. Check cluster status is 1 controller with 2 observers.

```
> ./bin/kafka-metadata-quorum.sh --bootstrap-controller localhost:9093 describe --status
ClusterId:              zuV6tFqAQhSGfFe97O-9cw
LeaderId:               1
LeaderEpoch:            1
HighWatermark:          817
MaxFollowerLag:         0
MaxFollowerLagTimeMs:   0
CurrentVoters:          [{"id": 1, "directoryId": "zD1nHI9OSyCGu29l3db2yg", "endpoints": ["CONTROLLER://localhost:9093"]}]
CurrentObservers:       [{"id": 2, "directoryId": "vX3tLI_VJKRLAMqJdFUY4Q"}, {"id": 3, "directoryId": "Mk1bbuDgirHiGzOnH4okew"}]
```

6. Add node 2 as controller and check cluster status is 2 controllers and 1 observer.

```
> ./bin/kafka-metadata-quorum.sh --bootstrap-controller localhost:9093 --command-config ./config/kraft/controller2.properties add-controller
Added controller 2 with directory id vX3tLI_VJKRLAMqJdFUY4Q and endpoints: CONTROLLER://localhost:9094
> ./bin/kafka-metadata-quorum.sh --bootstrap-controller localhost:9093 describe --status
ClusterId:              zuV6tFqAQhSGfFe97O-9cw
LeaderId:               1
LeaderEpoch:            1
HighWatermark:          960
MaxFollowerLag:         0
MaxFollowerLagTimeMs:   0
CurrentVoters:          [{"id": 1, "directoryId": "zD1nHI9OSyCGu29l3db2yg", "endpoints": ["CONTROLLER://localhost:9093"]}, {"id": 2, "directoryId": "vX3tLI_VJKRLAMqJdFUY4Q", "endpoints": ["CONTROLLER://localhost:9094"]}]
CurrentObservers:       [{"id": 3, "directoryId": "Mk1bbuDgirHiGzOnH4okew"}]
```

7. Add node 3 as controller and check cluster status is 3 controllers and 0 observer.

```
> ./bin/kafka-metadata-quorum.sh --bootstrap-controller localhost:9093 --command-config ./config/kraft/controller3.properties add-controller
Added controller 3 with directory id Mk1bbuDgirHiGzOnH4okew and endpoints: CONTROLLER://localhost:9095
> ./bin/kafka-metadata-quorum.sh --bootstrap-controller localhost:9093 describe --status
ClusterId:              zuV6tFqAQhSGfFe97O-9cw
LeaderId:               1
LeaderEpoch:            1
HighWatermark:          1123
MaxFollowerLag:         0
MaxFollowerLagTimeMs:   0
CurrentVoters:          [{"id": 1, "directoryId": "zD1nHI9OSyCGu29l3db2yg", "endpoints": ["CONTROLLER://localhost:9093"]}, {"id": 2, "directoryId": "vX3tLI_VJKRLAMqJdFUY4Q", "endpoints": ["CONTROLLER://localhost:9094"]}, {"id": 3, "directoryId": "Mk1bbuDgirHiGzOnH4okew", "endpoints": ["CONTROLLER://localhost:9095"]}]
CurrentObservers:       []
```

FrankYang0529 commented:
BTW, I found the adding controller example command is different from the KIP-853. Create a followup Jira: https://issues.apache.org/jira/browse/KAFKA-18059.

chia7712 commented:
> I found the adding controller command is different from the KIP-853. Create a followup Jira:

nice find and I have added feedback on the jira

chia7712 commented on core/src/test/scala/unit/kafka/tools/StorageToolTest.scala:
```diff
@@ -483,20 +483,44 @@ Found problem:
+  @Test
+  def testFormatWithNoInitialControllersFlagAndStandaloneFlagFails(): Unit = {
+    val arguments = ListBuffer[String](
+      "--release-version", "3.9-IV0",
```
please add "format" to match the test case

chia7712 commented on core/src/test/scala/unit/kafka/tools/StorageToolTest.scala:
```diff
@@ -483,20 +483,44 @@ Found problem:
+    val arguments = ListBuffer[String](
+      "--release-version", "3.9-IV0",
+      "--no-initial-controllers", "--standalone")
+    assertThrows(classOf[ArgumentParserException], () => StorageTool.parseArguments(arguments.toArray))
```
please check the exception message as well

chia7712 commented on core/src/test/scala/unit/kafka/tools/StorageToolTest.scala:
```diff
@@ -483,20 +483,44 @@ Found problem:
+  @Test
+  def testFormatWithNoInitialControllersFlagAndInitialControllersFlagFails(): Unit = {
+    val arguments = ListBuffer[String](
+      "--release-version", "3.9-IV0",
```
ditto

chia7712 reviewed (commented):
@FrankYang0529 thanks for your patch

FrankYang0529 committed:
KAFKA-18028: the effective kraft version of --no-initial-controllers should be 1 rather than 0

Signed-off-by: PoAn Yang <payang@apache.org>

FrankYang0529 commented:
Hi @chia7712, I addressed all comments. Thanks for the suggestion.

kevin-wu24 commented:
Hi @chia7712 @FrankYang0529, when a node sets `--no-initial-controllers`, the idea is that the node does not know what kraft.version the cluster should be running. 

Therefore, the flag should be a no-op, and the "effective kraft version" of a node setting this flag should be 0, not 1. The reason for this is because we can format a node with a old MV that does not support kraft.version=1. In this case, setting `--no-initial-controllers` causes formatting to fail, even though it should not.

Filed https://issues.apache.org/jira/browse/KAFKA-19719 to fix.
